flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/DeduplicateFunctionHelper.java

```diff
@@ -35,25 +37,29 @@
 	 *
 	 * @param currentRow latest row received by deduplicate function
 	 * @param generateUpdateBefore whether need to send UPDATE_BEFORE message for updates
-	 * @param state state of function
+	 * @param state state of function, null if generateUpdateBefore is false
 	 * @param out underlying collector
 	 */
 	static void processLastRow(
 			BaseRow currentRow,
 			boolean generateUpdateBefore,
-			ValueState<BaseRow> state,
+			@Nullable ValueState<BaseRow> state,
 			Collector<BaseRow> out) throws Exception {
-		// Check message should be accumulate
-		Preconditions.checkArgument(BaseRowUtil.isAccumulateMsg(currentRow));
+		// check message should be insert only.
+		Preconditions.checkArgument(currentRow.getRowKind() == RowKind.INSERT);
 		if (generateUpdateBefore) {
-			// state stores complete row if generateUpdateBefore is true
+			// state is not null when generateUpdateBefore is enabled,
+			// the state stores complete row
 			BaseRow preRow = state.value();
 			state.update(currentRow);
 			if (preRow != null) {
-				preRow.setHeader(BaseRowUtil.RETRACT_MSG);
+				preRow.setRowKind(RowKind.UPDATE_BEFORE);
 				out.collect(preRow);
 			}
 		}
+		// in order for better performance, we don't have state for LastRow
+		// if not generate UPDATE_BEFORE, thus, we can't produce INSERT messages for first row.
```

Comment:
Upsert databases uses `UPSERT` statement (or called REPLACE statement) which will replace values if the key exists, otherwise, it will insert a new row. So it works if all rows are `UPDATE`. 

Currently, the upsert sinks doesn't distinguish `INSERT` and `UPDATE` messages, they will be executed in `UPSERT` statements.

Reply:
does all database support this ? I know mysql supports "INSERT ON DUPLICATE KEY UPDATE"

Reply:
I changed the logic to emit INSERT for first row with the cost of additional state accessing.